### ISSUE TITLE
check should create cache directory if it doesn't exist instead of reporting an error

### ISSUE THREAD

**Issue by teran** [restic/restic#4437]
Output of `restic version`
--------------------------

It was 0.15.2, just checked on 0.16.0 - the same:

```
restic 0.16.0 compiled with go1.20.6 on darwin/amd64
```

How did you run restic exactly?
-------------------------------

```
TMPDIR=/Volumes/Flash/Temp/restic restic --repo repoPath  --password-command 'someCommand' --cache-dir /Volumes/Flash/Temp/Cache/restic/repo check
```

not critical error received:

```
unable to create temporary directory for cache during check, disabling cache: stat /Volumes/Flash/Temp/Cache/restic/repo: no such file or directory
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

local

Expected behavior
-----------------

cache directory will be created and will be used just like in other subcommands

Actual behavior
---------------

according to error message: cache is disabled, cache directory is not created

Steps to reproduce the behavior
-------------------------------

Run check on any repository with not existent but explicitly specified cache directory via flag

Do you have any idea what may have caused this?
-----------------------------------------------

haven't checked the source code in that place but it looks like different subcommands have different handling for the same flags

Do you have an idea how to solve the issue?
-------------------------------------------

If the guess above is right: just isolate flag handling and reuse it

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Sure, it works great and the issue is not critical at all :)

**Comment by MichaelEischer** [comment]
The underlying problem is that the temporary cache directory is created using `os.MkdirTemp`, which requires that the `cachedir` already exists. A `os.MkdirAll` before that, should do the trick. The alternative would be to integrate the tmp dir selection into the Cache, but that's probably not worth the effort.
 
https://github.com/restic/restic/blob/57a08291f5504c1f503957c9981f384cb35946cc/cmd/restic/cmd_check.go#L175

**Comment by teran** [comment]
> The underlying problem is that the temporary cache directory is created using `os.MkdirTemp`, which requires that the `cachedir` already exists. A `os.MkdirAll` before that, should do the trick. The alternative would be to integrate the tmp dir selection into the Cache, but that's probably not worth the effort.
> 
> https://github.com/restic/restic/blob/57a08291f5504c1f503957c9981f384cb35946cc/cmd/restic/cmd_check.go#L175

Sure, it will work.

My personal choice however would be something like that:


```
tempdirName, err := os.MkdirAll(path.Join(os.TempDir(), uuid.New().String()), 0700)
```

UUID (v4) here is just to get some entropy and since MkdirTemp makes no sense actually if you gonna create the whole path on your own - TempDir() anyway will always give you a proper path for that. But I'm a kinda paranoid guy :)

**Comment by plant99** [comment]
Hey, I'd love to work on this. @MichaelEischer could you please assign this to me?

PS: Been a very happy restic user for about a year now. I'd love to contribute code here on!

**Comment by MichaelEischer** [comment]
> os.TempDir()

@teran The OS temp dir is not always large enough to store the whole metadata of a repository. That's the whole reason why the cache directory is used if specified manually.

**Comment by plant99** [comment]
Apologies for the delay here folks, I've attempted a fix in the linked PR: https://github.com/restic/restic/pull/4805

**Comment by MichaelEischer** [comment]
Fixed by #4805
